### PR TITLE
Speed up rm bundles

### PR DESCRIPTION
**Pull request by AndrewJGaut**
### Reasons for making this change

Speed up rm bundles since it's very slow on the Stanford instance (and is slow in general on any instance that has lots of bundles).

### Related issues

#4217

### CODE REVIEW

**Comment by AndrewJGaut** [comment]
Speedup on dev for deleting bundles:
Original code: ~2.48 seconds
Updated code: ~0.243 seconds

**Comment by AndrewJGaut** [comment]
Note: the reason it was so slow prior is because it would recompute the user disk in entirety each time a bundle was removed (_get_disk_used(user_id) was called, which searches for all bundles owned by the user). To alleviate this, we now subtract the sum of the sizes of the deleted bundles from the user disk used instead, which is much faster.

File: codalab/model/bundle_model.py [hunk 1]
```diff
@@ -2708,6 +2708,14 @@ def increment_user_time_used(self, user_id, amount):
         user_info['time_used'] += amount
         self.update_user_info(user_info)
 
+    def increment_user_disk_used(self, user_id, amount):
```
Comment (percyliang) [inline codalab/model/bundle_model.py]: Add type hints.

File: codalab/rest/bundles.py [hunk 2]
```diff
@@ -1314,7 +1315,12 @@ def delete_bundles(uuids, force, recursive, data_only, dry_run):
             local.model.delete_bundles(relevant_uuids)
 
         # Update user statistics
-        local.model.update_user_disk_used(request.user.user_id)
+        # Just decrement the user disk used by the sum of sizes of bundles deleted
+        # OK, now let's add our change.
+        bundle_data_sizes = local.model.get_bundle_metadata(relevant_uuids, 'data_size')
+        local.model.increment_user_disk_used(
+            request.user.user_id, (-1) * sum(map(int, bundle_data_sizes.values()))
```
Comment (percyliang) [inline codalab/rest/bundles.py]: Can just do `-sum...`

File: codalab/rest/bundles.py [hunk 3]
```diff
@@ -1314,7 +1315,12 @@ def delete_bundles(uuids, force, recursive, data_only, dry_run):
             local.model.delete_bundles(relevant_uuids)
 
         # Update user statistics
-        local.model.update_user_disk_used(request.user.user_id)
+        # Just decrement the user disk used by the sum of sizes of bundles deleted
+        # OK, now let's add our change.
+        bundle_data_sizes = local.model.get_bundle_metadata(relevant_uuids, 'data_size')
```
Comment (percyliang) [inline codalab/rest/bundles.py]: Does this diff produce the same answer as the above?  Just wondering if there are subtle differences?

File: codalab/rest/bundles.py [hunk 4]
```diff
@@ -1314,7 +1315,12 @@ def delete_bundles(uuids, force, recursive, data_only, dry_run):
             local.model.delete_bundles(relevant_uuids)
 
         # Update user statistics
-        local.model.update_user_disk_used(request.user.user_id)
+        # Just decrement the user disk used by the sum of sizes of bundles deleted
+        # OK, now let's add our change.
```
Comment (percyliang) [inline codalab/rest/bundles.py]: Delete this comment line?

**Review by percyliang — COMMENTED** [review]
Ah, very nice!  Wonder if we could add a test for this.

File: codalab/rest/bundles.py (hunk 3 above)
Comment (percyliang) [inline codalab/rest/bundles.py]: Also check the case where some bundles that are being deleted don't have data (e.g., a bundle which has had `cl rm -d ...` done on it)

File: codalab/lib/path_util.py [hunk 5]
```diff
@@ -310,20 +310,24 @@ def remove(path):
 
         if not FileSystems.exists(path):
             FileSystems.delete([path])
-        return
+        return True  # not sure about this one
```
Comment (wwwjn) [inline codalab/lib/path_util.py]: `return True` look good to me

File: codalab/lib/path_util.py [hunk 6]
```diff
@@ -310,20 +310,24 @@ def remove(path):
...
     check_isvalid(path, 'remove')
     set_write_permissions(path)  # Allow permissions
     if os.path.islink(path):
         os.unlink(path)
+        return False
     elif os.path.isdir(path):
         try:
             shutil.rmtree(path)
+            return True
         except shutil.Error:
```
Comment (wwwjn) [inline codalab/lib/path_util.py]: If there is an error, it will have no return value here? Should we use `finally` here?

File: codalab/model/bundle_model.py (hunk 1 above)
Comment (AndrewJGaut) [reply codalab/model/bundle_model.py]: Done.

File: codalab/rest/bundles.py (hunk 2 above)
Comment (AndrewJGaut) [reply codalab/rest/bundles.py]: Modified this to use something different anyways

File: codalab/rest/bundles.py (hunk 3 above)
Comment (AndrewJGaut) [reply codalab/rest/bundles.py]: It did not, it turns out, due to the following edge cases:
- cl rm -d: Here, the bundle contents are removed but not the bundle metadata. For a given bundle B, we must take care to decrement the user disk quota when *cl rm -d B* is run, but to not decrement it again when *cl rm B* is run (since that second *cl rm* just deletes the metadata and not anything from disk).
- Files uploaded with symlinks: These shouldn't affect user disk quota regardless, so when we remove them there should be no decrement to the user disk quota.

The new changes fix this.

File: codalab/rest/bundles.py (hunk 4 above)
Comment (AndrewJGaut) [reply codalab/rest/bundles.py]: Done.

File: codalab/rest/bundles.py (hunk 3 above)
Comment (AndrewJGaut) [reply codalab/rest/bundles.py]: Checked that case with the new tests.

File: codalab/lib/path_util.py (hunk 6 above)
Comment (AndrewJGaut) [reply codalab/lib/path_util.py]: Good catch! I added a return False statement at the end of the function to catch any other cases.

**Comment by AndrewJGaut** [comment]
I made some changes to catch edge cases.

Previously, there were no test cases that I saw for verifying that disk quota was incremented and decremented properly upon file upload and deletion. So, I added test cases to check that disk quota is incremented and decremented properly upon:
- normal file upload and deletion
- normal file upload and deletion with rm -d before deletion with rm
- symlinked file upload and deletion

This helps to verify that my rm modification still runs correctly and to verify that our code handles disk quota usage correctly in various cases.

**Comment by AndrewJGaut** [comment]
Now ready for final review @percyliang @wwwjn

Thanks for the reviews!

**Review by wwwjn — APPROVED** [review]
